src/rest_vol_dataset.c

```diff
@@ -416,207 +438,338 @@ RV_dataset_open(void *obj, const H5VL_loc_params_t *loc_params, const char *name
  *              March, 2017
  */
 herr_t
-RV_dataset_read(size_t count, void *dset[], hid_t mem_type_id[], hid_t mem_space_id[], hid_t file_space_id[],
-                hid_t dxpl_id, void *buf[], void **req)
+RV_dataset_read(size_t count, void *dset[], hid_t mem_type_id[], hid_t _mem_space_id[],
+                hid_t _file_space_id[], hid_t dxpl_id, void *buf[], void **req)
```

Comment:
Just an observation: with the large amount of things you're allocating in this function, the multi dataset path may be suffering from a lot of cache misses as you jump between the ~19-21 arrays allocated here. I believe it will be better for performance and much more readable to declare a structure type that has a field for everything you need related to an individual datasets and then allocate and free a `count`-sized array of those structures instead of having every array be separate in memory.